// dappscout-iframe/dist/dappscout-iframe.es.js
var gr = Object.defineProperty;
var hr = (o, n, c) => n in o ? gr(o, n, { enumerable: !0, configurable: !0, writable: !0, value: c }) : o[n] = c;
var O = (o, n, c) => (hr(o, typeof n != "symbol" ? n + "" : n, c), c);
import Le, { useState as se, useEffect as ue, createContext as Er, useRef as yr, useCallback as ie, useContext as Rr } from "react";
import { providers as mr, utils as _r } from "ethers";
var le = { exports: {} }, Y = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Fe;
function br() {
  if (Fe)
    return Y;
  Fe = 1;
  var o = Le, n = Symbol.for("react.element"), c = Symbol.for("react.fragment"), v = Object.prototype.hasOwnProperty, y = o.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, R = { key: !0, ref: !0, __self: !0, __source: !0 };
  function T(_, m, k) {
    var b, C = {}, i = null, x = null;
    k !== void 0 && (i = "" + k), m.key !== void 0 && (i = "" + m.key), m.ref !== void 0 && (x = m.ref);
    for (b in m)
      v.call(m, b) && !R.hasOwnProperty(b) && (C[b] = m[b]);
    if (_ && _.defaultProps)
      for (b in m = _.defaultProps, m)
        C[b] === void 0 && (C[b] = m[b]);
    return { $$typeof: n, type: _, key: i, ref: x, props: C, _owner: y.current };
  }
  return Y.Fragment = c, Y.jsx = T, Y.jsxs = T, Y;
}
var B = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var We;
function Cr() {
  return We || (We = 1, process.env.NODE_ENV !== "production" && function() {
    var o = Le, n = Symbol.for("react.element"), c = Symbol.for("react.portal"), v = Symbol.for("react.fragment"), y = Symbol.for("react.strict_mode"), R = Symbol.for("react.profiler"), T = Symbol.for("react.provider"), _ = Symbol.for("react.context"), m = Symbol.for("react.forward_ref"), k = Symbol.for("react.suspense"), b = Symbol.for("react.suspense_list"), C = Symbol.for("react.memo"), i = Symbol.for("react.lazy"), x = Symbol.for("react.offscreen"), N = Symbol.iterator, W = "@@iterator";
    function f(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = N && e[N] || e[W];
      return typeof r == "function" ? r : null;
    }
    var p = o.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function d(e) {
      {
        for (var r = arguments.length, t = new Array(r > 1 ? r - 1 : 0), a = 1; a < r; a++)
          t[a - 1] = arguments[a];
        j("error", e, t);
      }
    }
    function j(e, r, t) {
      {
        var a = p.ReactDebugCurrentFrame, l = a.getStackAddendum();
        l !== "" && (r += "%s", t = t.concat([l]));
        var g = t.map(function(u) {
          return String(u);
        });
        g.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, g);
      }
    }
    var D = !1, K = !1, X = !1, fe = !1, Ve = !1, de;
    de = Symbol.for("react.module.reference");
    function Ye(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === v || e === R || Ve || e === y || e === k || e === b || fe || e === x || D || K || X || typeof e == "object" && e !== null && (e.$$typeof === i || e.$$typeof === C || e.$$typeof === T || e.$$typeof === _ || e.$$typeof === m || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === de || e.getModuleId !== void 0));
    }
    function Be(e, r, t) {
      var a = e.displayName;
      if (a)
        return a;
      var l = r.displayName || r.name || "";
      return l !== "" ? t + "(" + l + ")" : t;
    }
    function ve(e) {
      return e.displayName || "Context";
    }
    function P(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && d("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case v:
          return "Fragment";
        case c:
          return "Portal";
        case R:
          return "Profiler";
        case y:
          return "StrictMode";
        case k:
          return "Suspense";
        case b:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case _:
            var r = e;
            return ve(r) + ".Consumer";
          case T:
            var t = e;
            return ve(t._context) + ".Provider";
          case m:
            return Be(e, e.render, "ForwardRef");
          case C:
            var a = e.displayName || null;
            return a !== null ? a : P(e.type) || "Memo";
          case i: {
            var l = e, g = l._payload, u = l._init;
            try {
              return P(u(g));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var F = Object.assign, M = 0, pe, ge, he, Ee, ye, Re, me;
    function _e() {
    }
    _e.__reactDisabledLog = !0;
    function Ue() {
      {
        if (M === 0) {
          pe = console.log, ge = console.info, he = console.warn, Ee = console.error, ye = console.group, Re = console.groupCollapsed, me = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: _e,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        M++;
      }
    }
    function qe() {
      {
        if (M--, M === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: F({}, e, {
              value: pe
            }),
            info: F({}, e, {
              value: ge
            }),
            warn: F({}, e, {
              value: he
            }),
            error: F({}, e, {
              value: Ee
            }),
            group: F({}, e, {
              value: ye
            }),
            groupCollapsed: F({}, e, {
              value: Re
            }),
            groupEnd: F({}, e, {
              value: me
            })
          });
        }
        M < 0 && d("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var Z = p.ReactCurrentDispatcher, Q;
    function q(e, r, t) {
      {
        if (Q === void 0)
          try {
            throw Error();
          } catch (l) {
            var a = l.stack.trim().match(/\n( *(at )?)/);
            Q = a && a[1] || "";
          }
        return `
` + Q + e;
      }
    }
    var ee = !1, G;
    {
      var Ge = typeof WeakMap == "function" ? WeakMap : Map;
      G = new Ge();
    }
    function be(e, r) {
      if (!e || ee)
        return "";
      {
        var t = G.get(e);
        if (t !== void 0)
          return t;
      }
      var a;
      ee = !0;
      var l = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var g;
      g = Z.current, Z.current = null, Ue();
      try {
        if (r) {
          var u = function() {
            throw Error();
          };
          if (Object.defineProperty(u.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(u, []);
            } catch (I) {
              a = I;
            }
            Reflect.construct(e, [], u);
          } else {
            try {
              u.call();
            } catch (I) {
              a = I;
            }
            e.call(u.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (I) {
            a = I;
          }
          e();
        }
      } catch (I) {
        if (I && a && typeof I.stack == "string") {
          for (var s = I.stack.split(`
`), w = a.stack.split(`
`), h = s.length - 1, E = w.length - 1; h >= 1 && E >= 0 && s[h] !== w[E]; )
            E--;
          for (; h >= 1 && E >= 0; h--, E--)
            if (s[h] !== w[E]) {
              if (h !== 1 || E !== 1)
                do
                  if (h--, E--, E < 0 || s[h] !== w[E]) {
                    var S = `
` + s[h].replace(" at new ", " at ");
                    return e.displayName && S.includes("<anonymous>") && (S = S.replace("<anonymous>", e.displayName)), typeof e == "function" && G.set(e, S), S;
                  }
                while (h >= 1 && E >= 0);
              break;
            }
        }
      } finally {
        ee = !1, Z.current = g, qe(), Error.prepareStackTrace = l;
      }
      var $ = e ? e.displayName || e.name : "", Ne = $ ? q($) : "";
      return typeof e == "function" && G.set(e, Ne), Ne;
    }
    function He(e, r, t) {
      return be(e, !1);
    }
    function Je(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function H(e, r, t) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return be(e, Je(e));
      if (typeof e == "string")
        return q(e);
      switch (e) {
        case k:
          return q("Suspense");
        case b:
          return q("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case m:
            return He(e.render);
          case C:
            return H(e.type, r, t);
          case i: {
            var a = e, l = a._payload, g = a._init;
            try {
              return H(g(l), r, t);
            } catch {
            }
          }
        }
      return "";
    }
    var J = Object.prototype.hasOwnProperty, Ce = {}, we = p.ReactDebugCurrentFrame;
    function z(e) {
      if (e) {
        var r = e._owner, t = H(e.type, e._source, r ? r.type : null);
        we.setExtraStackFrame(t);
      } else
        we.setExtraStackFrame(null);
    }
    function ze(e, r, t, a, l) {
      {
        var g = Function.call.bind(J);
        for (var u in e)
          if (g(e, u)) {
            var s = void 0;
            try {
              if (typeof e[u] != "function") {
                var w = Error((a || "React class") + ": " + t + " type `" + u + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[u] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw w.name = "Invariant Violation", w;
              }
              s = e[u](r, u, a, t, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (h) {
              s = h;
            }
            s && !(s instanceof Error) && (z(l), d("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", a || "React class", t, u, typeof s), z(null)), s instanceof Error && !(s.message in Ce) && (Ce[s.message] = !0, z(l), d("Failed %s type: %s", t, s.message), z(null));
          }
      }
    }
    var Ke = Array.isArray;
    function re(e) {
      return Ke(e);
    }
    function Xe(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, t = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return t;
      }
    }
    function Ze(e) {
      try {
        return Te(e), !1;
      } catch {
        return !0;
      }
    }
    function Te(e) {
      return "" + e;
    }
    function Se(e) {
      if (Ze(e))
        return d("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", Xe(e)), Te(e);
    }
    var V = p.ReactCurrentOwner, Qe = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, Oe, xe, te;
    te = {};
    function er(e) {
      if (J.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function rr(e) {
      if (J.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function tr(e, r) {
      if (typeof e.ref == "string" && V.current && r && V.current.stateNode !== r) {
        var t = P(V.current.type);
        te[t] || (d('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', P(V.current.type), e.ref), te[t] = !0);
      }
    }
    function nr(e, r) {
      {
        var t = function() {
          Oe || (Oe = !0, d("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: t,
          configurable: !0
        });
      }
    }
    function ar(e, r) {
      {
        var t = function() {
          xe || (xe = !0, d("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: t,
          configurable: !0
        });
      }
    }
    var or = function(e, r, t, a, l, g, u) {
      var s = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: n,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: t,
        props: u,
        // Record the component responsible for creating this element.
        _owner: g
      };
      return s._store = {}, Object.defineProperty(s._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(s, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: a
      }), Object.defineProperty(s, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: l
      }), Object.freeze && (Object.freeze(s.props), Object.freeze(s)), s;
    };
    function ir(e, r, t, a, l) {
      {
        var g, u = {}, s = null, w = null;
        t !== void 0 && (Se(t), s = "" + t), rr(r) && (Se(r.key), s = "" + r.key), er(r) && (w = r.ref, tr(r, l));
        for (g in r)
          J.call(r, g) && !Qe.hasOwnProperty(g) && (u[g] = r[g]);
        if (e && e.defaultProps) {
          var h = e.defaultProps;
          for (g in h)
            u[g] === void 0 && (u[g] = h[g]);
        }
        if (s || w) {
          var E = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          s && nr(u, E), w && ar(u, E);
        }
        return or(e, s, w, l, a, V.current, u);
      }
    }
    var ne = p.ReactCurrentOwner, Pe = p.ReactDebugCurrentFrame;
    function L(e) {
      if (e) {
        var r = e._owner, t = H(e.type, e._source, r ? r.type : null);
        Pe.setExtraStackFrame(t);
      } else
        Pe.setExtraStackFrame(null);
    }
    var ae;
    ae = !1;
    function oe(e) {
      return typeof e == "object" && e !== null && e.$$typeof === n;
    }
    function Ie() {
      {
        if (ne.current) {
          var e = P(ne.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function sr(e) {
      {
        if (e !== void 0) {
          var r = e.fileName.replace(/^.*[\\\/]/, ""), t = e.lineNumber;
          return `

Check your code at ` + r + ":" + t + ".";
        }
        return "";
      }
    }
    var Ae = {};
    function ur(e) {
      {
        var r = Ie();
        if (!r) {
          var t = typeof e == "string" ? e : e.displayName || e.name;
          t && (r = `

Check the top-level render call using <` + t + ">.");
        }
        return r;
      }
    }
    function ke(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var t = ur(r);
        if (Ae[t])
          return;
        Ae[t] = !0;
        var a = "";
        e && e._owner && e._owner !== ne.current && (a = " It was passed a child from " + P(e._owner.type) + "."), L(e), d('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', t, a), L(null);
      }
    }
    function je(e, r) {
      {
        if (typeof e != "object")
          return;
        if (re(e))
          for (var t = 0; t < e.length; t++) {
            var a = e[t];
            oe(a) && ke(a, r);
          }
        else if (oe(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var l = f(e);
          if (typeof l == "function" && l !== e.entries)
            for (var g = l.call(e), u; !(u = g.next()).done; )
              oe(u.value) && ke(u.value, r);
        }
      }
    }
    function lr(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var t;
        if (typeof r == "function")
          t = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === m || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === C))
          t = r.propTypes;
        else
          return;
        if (t) {
          var a = P(r);
          ze(t, e.props, "prop", a, e);
        } else if (r.PropTypes !== void 0 && !ae) {
          ae = !0;
          var l = P(r);
          d("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", l || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && d("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function cr(e) {
      {
        for (var r = Object.keys(e.props), t = 0; t < r.length; t++) {
          var a = r[t];
          if (a !== "children" && a !== "key") {
            L(e), d("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", a), L(null);
            break;
          }
        }
        e.ref !== null && (L(e), d("Invalid attribute `ref` supplied to `React.Fragment`."), L(null));
      }
    }
    function De(e, r, t, a, l, g) {
      {
        var u = Ye(e);
        if (!u) {
          var s = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (s += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var w = sr(l);
          w ? s += w : s += Ie();
          var h;
          e === null ? h = "null" : re(e) ? h = "array" : e !== void 0 && e.$$typeof === n ? (h = "<" + (P(e.type) || "Unknown") + " />", s = " Did you accidentally export a JSX literal instead of a component?") : h = typeof e, d("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", h, s);
        }
        var E = ir(e, r, t, l, g);
        if (E == null)
          return E;
        if (u) {
          var S = r.children;
          if (S !== void 0)
            if (a)
              if (re(S)) {
                for (var $ = 0; $ < S.length; $++)
                  je(S[$], e);
                Object.freeze && Object.freeze(S);
              } else
                d("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              je(S, e);
        }
        return e === v ? cr(E) : lr(E), E;
      }
    }
    function fr(e, r, t) {
      return De(e, r, t, !0);
    }
    function dr(e, r, t) {
      return De(e, r, t, !1);
    }
    var vr = dr, pr = fr;
    B.Fragment = v, B.jsx = vr, B.jsxs = pr;
  }()), B;
}
process.env.NODE_ENV === "production" ? le.exports = br() : le.exports = Cr();
var wr = le.exports;
const $e = () => "7.6.0", Tr = (o) => o.toString(16).padStart(2, "0"), Sr = (o) => {
  const n = new Uint8Array((o || 40) / 2);
  return window.crypto.getRandomValues(n), Array.from(n, Tr).join("");
}, Or = () => typeof window < "u" ? Sr(10) : (/* @__PURE__ */ new Date()).getTime().toString(36);
class U {
}
O(U, "makeRequest", (n, c) => ({
  id: Or(),
  method: n,
  params: c,
  env: {
    sdkVersion: $e()
  }
})), O(U, "makeResponse", (n, c, v) => ({
  id: n,
  success: !0,
  version: v,
  data: c
})), O(U, "makeErrorResponse", (n, c, v) => ({
  id: n,
  success: !1,
  error: c,
  version: v
}));
var A = /* @__PURE__ */ ((o) => (o.sendTransactions = "sendTransactions", o.rpcCall = "rpcCall", o.getChainInfo = "getChainInfo", o.getSafeInfo = "getSafeInfo", o.getTxBySafeTxHash = "getTxBySafeTxHash", o.getSafeBalances = "getSafeBalances", o.signMessage = "signMessage", o.signTypedMessage = "signTypedMessage", o.getEnvironmentInfo = "getEnvironmentInfo", o.requestAddressBook = "requestAddressBook", o.wallet_getPermissions = "wallet_getPermissions", o.wallet_requestPermissions = "wallet_requestPermissions", o))(A || {}), ce = /* @__PURE__ */ ((o) => (o.AWAITING_CONFIRMATIONS = "AWAITING_CONFIRMATIONS", o.AWAITING_EXECUTION = "AWAITING_EXECUTION", o.CANCELLED = "CANCELLED", o.FAILED = "FAILED", o.SUCCESS = "SUCCESS", o.PENDING = "PENDING", o.WILL_BE_REPLACED = "WILL_BE_REPLACED", o))(ce || {});
class xr {
  constructor(n) {
    O(this, "iframeRef");
    O(this, "handlers", /* @__PURE__ */ new Map());
    O(this, "on", (n, c) => {
      this.handlers.set(n, c);
    });
    O(this, "isValidMessage", (n) => {
      var y;
      if (Object.prototype.hasOwnProperty.call(n.data, "isCookieEnabled"))
        return !0;
      const c = ((y = this.iframeRef.current) == null ? void 0 : y.contentWindow) === n.source, v = Object.values(A).includes(n.data.method);
      return c && v;
    });
    O(this, "canHandleMessage", (n) => !!this.handlers.get(n.data.method));
    O(this, "send", (n, c, v = !1) => {
      var T, _;
      const y = $e(), R = v ? U.makeErrorResponse(
        c,
        n,
        y
      ) : U.makeResponse(c, n, y);
      (_ = (T = this.iframeRef.current) == null ? void 0 : T.contentWindow) == null || _.postMessage(R, "*");
    });
    O(this, "handleIncomingMessage", async (n) => {
      const c = this.isValidMessage(n), v = this.canHandleMessage(n);
      if (c && v) {
        n.data.method === "rpcCall" && // eslint-disable-next-line @typescript-eslint/no-explicit-any
        n.data.params.call === "eth_getBlockByNumber" || console.log("*", n.data);
        const y = this.handlers.get(n.data.method);
        try {
          const R = await y(n);
          typeof R < "u" && this.send(R, n.data.id);
        } catch (R) {
          this.send(R.message, n.data.id, !0);
        }
      }
    });
    O(this, "clear", () => {
      window.removeEventListener("message", this.handleIncomingMessage);
    });
    this.iframeRef = n, window.addEventListener("message", this.handleIncomingMessage);
  }
}
const Pr = (o) => {
  const [n, c] = se(
    void 0
  );
  return ue(() => {
    let v;
    return ((R) => {
      v = new xr(R), c(v);
    })(o), () => {
      v == null || v.clear();
    };
  }, [o]), n;
}, Me = Er({
  iframeRef: null,
  isReady: !1
}), jr = ({
  children: o,
  address: n,
  appUrl: c,
  rpcUrl: v,
  sendTransaction: y,
  signMessage: R,
  signTypedData: T
}) => {
  const [_, m] = se(), [k, b] = se(!1), C = yr(null), i = Pr(C), x = ie(
    function(f, p) {
      var j, D;
      const d = {
        ...f,
        requestId: p || Math.trunc(window.performance.now()),
        version: "0.4.2"
      };
      C && ((D = (j = C.current) == null ? void 0 : j.contentWindow) == null || D.postMessage(
        d,
        c
      ));
    },
    [C, c]
  ), N = ie((f, p) => {
    x({ messageId: "TRANSACTION_CONFIRMED", data: f }, p), i == null || i.send(f, p);
  }, [i, x]), W = ie((f) => {
    x({ messageId: "TRANSACTION_REJECTED", data: {} }, f), i == null || i.send("Transaction was rejected", f, !0);
  }, [i, x]);
  return ue(() => {
    v && m(new mr.StaticJsonRpcProvider(v));
  }, [v]), ue(() => {
    _ && (i == null || i.on(A.getSafeInfo, async () => ({
      safeAddress: n,
      chainId: (await _.getNetwork()).chainId,
      owners: [],
      threshold: 1,
      isReadOnly: !1
    })), i == null || i.on(A.getEnvironmentInfo, async () => ({
      origin: document.location.origin
    })), i == null || i.on(A.rpcCall, async (f) => {
      const p = f.data.params;
      try {
        return await _.send(
          p.call,
          p.params
        );
      } catch (d) {
        return d;
      }
    }), i == null || i.on(A.getTxBySafeTxHash, async (f) => {
      console.log("communicator.getTxBySafeTxHash", f);
      const { safeTxHash: p } = f.data.params, { timestamp: d, blockNumber: j, to: D, data: K, value: X } = await _.getTransaction(p);
      return {
        txId: p,
        txStatus: j ? ce.SUCCESS : ce.PENDING,
        executedAt: d,
        txInfo: {
          type: "Custom",
          to: {
            value: D ?? ""
          },
          dataSize: K,
          value: X.toString(),
          isCancellation: !1
        },
        txHash: p
      };
    }), i == null || i.on(A.sendTransactions, async (f) => {
      console.log("communicator.sendTransactions", f);
      try {
        const p = f.data.params.txs.map(
          ({ to: j, ...D }) => ({
            to: _r.getAddress(j),
            // checksummed
            ...D
          })
        ), d = await (y == null ? void 0 : y(p[0]));
        N({ safeTxHash: d }, f.data.id);
      } catch {
        W(f.data.id);
      }
    }), i == null || i.on(A.signMessage, async (f) => {
      console.log("communicator.signMessage", f);
      try {
        const { message: p } = f.data.params, d = await (R == null ? void 0 : R(p));
        N({ signature: d }, f.data.id);
      } catch {
        W(f.data.id);
      }
    }), i == null || i.on(A.signTypedMessage, async (f) => {
      console.log("communicator.signTypedMessage", f);
      try {
        const { typedData: p } = f.data.params, d = await (T == null ? void 0 : T(p));
        N({ signature: d }, f.data.id);
      } catch {
        W(f.data.id);
      }
    }), b(!0));
  }, [
    i,
    n,
    _,
    N,
    W,
    y,
    R,
    T
  ]), /* @__PURE__ */ wr.jsx(Me.Provider, { value: { iframeRef: C, isReady: k }, children: o });
}, Dr = () => Rr(Me);
export {
  jr as DappscoutIframeProvider,
  Dr as useDappscoutIframe
};
