// @growthbook/growthbook/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GrowthBook", {
  enumerable: true,
  get: function () {
    return _GrowthBook.GrowthBook;
  }
});
Object.defineProperty(exports, "clearCache", {
  enumerable: true,
  get: function () {
    return _featureRepository.clearCache;
  }
});
Object.defineProperty(exports, "configureCache", {
  enumerable: true,
  get: function () {
    return _featureRepository.configureCache;
  }
});
Object.defineProperty(exports, "helpers", {
  enumerable: true,
  get: function () {
    return _featureRepository.helpers;
  }
});
Object.defineProperty(exports, "isURLTargeted", {
  enumerable: true,
  get: function () {
    return _util.isURLTargeted;
  }
});
Object.defineProperty(exports, "onHidden", {
  enumerable: true,
  get: function () {
    return _featureRepository.onHidden;
  }
});
Object.defineProperty(exports, "onVisible", {
  enumerable: true,
  get: function () {
    return _featureRepository.onVisible;
  }
});
Object.defineProperty(exports, "setPolyfills", {
  enumerable: true,
  get: function () {
    return _featureRepository.setPolyfills;
  }
});
var _featureRepository = require("./feature-repository");
var _GrowthBook = require("./GrowthBook");
var _util = require("./util");
